internal/provider/resource_machine.go

```diff
@@ -571,3 +620,21 @@ func updateAnnotations(ctx context.Context, client annotationSetter, stateAnnota
 	}
 	return diagnostics
 }
+
+func assertHostnamePopulated(respFromAPI juju.ReadMachineResponse) error {
+	if respFromAPI.Hostname == "" {
+		return juju.NewRetryReadError("waiting for hostname to be set on machine")
+	}
+	return nil
+}
+
+// assertMachineRunning waits for machine to go into "running" state. This is important when using the placement directive
```

Comment:
assertMachineRunning asserts that the machine is in the `running` state, otherwise it returns a retry error.